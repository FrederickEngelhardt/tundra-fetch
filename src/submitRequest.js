export default async (capturedRequest, port, matched = true) => {
  // Using XMLHttpRequest to not interfere with the overwritten fetch object.
  // Sending in a fire-and-forget fashion.
  const request = new XMLHttpRequest(); // eslint-disable-line no-undef
  const requestUrl = `http://localhost:${port}/${matched ? 'requests' : 'requests/type/unmatched'}`;

  request.open('POST', requestUrl);
  request.setRequestHeader('Content-Type', 'application/json');
  request.send(JSON.stringify(capturedRequest));
};
