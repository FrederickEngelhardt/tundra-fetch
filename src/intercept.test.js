import fetchMock from 'fetch-mock';
import interceptFetchCalls from './intercept';
import submitRequest from './submitRequest';

jest.mock('./submitRequest', () => jest.fn());

describe('intercept', () => {
  const callback = jest.fn();

  beforeEach(() => {
    fetchMock.restore();
    jest.clearAllMocks();
  });

  it('should intercept a plain url', async () => {
    fetchMock.get('http://someurl.com', { data: 'abc123' });
    interceptFetchCalls(12345, callback);

    const response = await (await global.fetch('http://someurl.com')).json();

    expect(response).toEqual({ data: 'abc123' });
    expect(submitRequest.mock.calls).toMatchSnapshot();
    expect(callback).toHaveBeenCalled();
  });

  it('should intercept a url with custom options', async () => {
    const headers = { some: 'header' };
    const options = {
      method: 'GET',
      headers,
    };

    fetchMock.get('http://someurl.com', { data: 'abc123' }, { headers });
    interceptFetchCalls(12345, callback);

    const response = await (await global.fetch('http://someurl.com', options)).json();

    expect(response).toEqual({ data: 'abc123' });
    expect(submitRequest.mock.calls).toMatchSnapshot();
    expect(callback).toHaveBeenCalled();
  });

  it('should intercept a POST fetch with a json body', async () => {
    const body = { some: 'request body' };
    const headers = { some: 'header' };
    const interceptedCalls = [];

    fetchMock.post('http://someurl.com', { data: 'abc123' }, { headers, body });
    interceptFetchCalls(12345, request => interceptedCalls.push(request));

    const options = {
      method: 'POST',
      headers,
      body,
    };

    const response = await (await global.fetch('http://someurl.com', options)).json();

    expect(response).toEqual({ data: 'abc123' });
    expect(submitRequest.mock.calls).toMatchSnapshot();

    expect(interceptedCalls.length).toBe(1);
    expect(interceptedCalls[0]).toEqual({
      request: {
        url: 'http://someurl.com',
        headers: { some: 'header' },
        method: 'POST',
        content: { some: 'request body' },
      },
      response: {
        headers: undefined,
        statusCode: 200,
        content: { data: 'abc123' },
      },
    });
  });

  it('should log an error to the console when there is a problem submitting request data', async () => {
    fetchMock.get('http://someurl.com', { data: 'abc123' });
    interceptFetchCalls(12345);
    submitRequest.mockImplementation(() => { throw new Error(); });
    global.console = { error: jest.fn() };

    const response = await (await global.fetch('http://someurl.com')).json();

    expect(response).toEqual({ data: 'abc123' });
    expect(global.console.error.mock.calls).toMatchSnapshot();
    expect(callback).not.toHaveBeenCalled();
  });
});
