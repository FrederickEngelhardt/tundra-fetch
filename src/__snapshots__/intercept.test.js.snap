// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`intercept should intercept a POST fetch with a json body 1`] = `
Array [
  Array [
    Object {
      "request": Object {
        "content": Object {
          "some": "request body",
        },
        "headers": Object {
          "some": "header",
        },
        "method": "POST",
        "url": "http://someurl.com",
      },
      "response": Object {
        "content": Object {
          "data": "abc123",
        },
        "headers": undefined,
        "statusCode": 200,
      },
    },
    12345,
  ],
]
`;

exports[`intercept should intercept a plain url 1`] = `
Array [
  Array [
    Object {
      "request": Object {
        "content": undefined,
        "headers": undefined,
        "method": "GET",
        "url": "http://someurl.com",
      },
      "response": Object {
        "content": Object {
          "data": "abc123",
        },
        "headers": undefined,
        "statusCode": 200,
      },
    },
    12345,
  ],
]
`;

exports[`intercept should intercept a url with custom options 1`] = `
Array [
  Array [
    Object {
      "request": Object {
        "content": undefined,
        "headers": Object {
          "some": "header",
        },
        "method": "GET",
        "url": "http://someurl.com",
      },
      "response": Object {
        "content": Object {
          "data": "abc123",
        },
        "headers": undefined,
        "statusCode": 200,
      },
    },
    12345,
  ],
]
`;

exports[`intercept should log an error to the console when there is a problem submitting request data 1`] = `
Array [
  Array [
    "Error wiretapping fetch request",
  ],
  Array [
    [Error],
  ],
]
`;
